fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,123 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+    fetch('/api/todos')
+      .then(response => response.json())
+      .then(data => this.setState({ todos: data }))
+      .catch(err => console.error('Fetch failed!', err));
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    *   - create a shallow copy of the todos array from state
+    *   - add the todo received from the server to the copied array
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(newTodo => {
+        const copy = this.state.todos;
```

Comment:
This doesn't create a copy of the array, just a reference to the same array. Since you are using `concat` on the next line it doesn't matter in this case, but important to understand that this will not create a copy of the array.